[Automation] java run fails for com.google.inject.extensions:guice-servlet:4.0

## java run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/26036219276/job/76536247556
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
uice-servlet:3.0:4.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 4.0 at 2026-05-18T13:30:58Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean compileTestJava -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 1s
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :classes UP-TO-DATE
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java uses or overrides a deprecated API.

Note: Recompile with -Xlint:deprecation for details.
> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 1s
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 3s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 4.0 at 2026-05-18T13:31:01Z with exit code 0 after 3s
========================================================================================
 com.google.inject.extensions:guice-servlet:3.0:4.0 stage java run
========================================================================================
Starting stage 'java run' for 4.0 at 2026-05-18T13:31:01Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 916ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :processTestResources NO-SOURCE

> Task :compileTestJava
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

> Task :testClasses

> Task :test FAILED

Guice_servletTest > duplicateServletMappingsAreReportedByServletModuleElements() FAILED
    java.lang.ExceptionInInitializerError
        at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
        at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
        at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
        at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
        at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
        at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
        at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
        at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
        at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.spi.Elements.getElements(Elements.java:87)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.duplicateServletMappingsAreReportedByServletModuleElements(Guice_servletTest.java:141)

        Caused by:
        java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6b5966e1
            at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
            at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:203)
            at java.base/java.lang.reflect.Method.setAccessible(Method.java:197)
            at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
            at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
            at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
            ... 20 more

Guice_servletTest > servletModuleExposesLinkedAndInstanceBindingsWithMatchingRules() FAILED
    java.lang.NoClassDefFoundError: Could not initialize class com.google.inject.internal.cglib.core.$ReflectUtils
        at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
        at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
        at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
        at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
        at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
        at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
        at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
        at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
        at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.spi.Elements.getElements(Elements.java:87)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.servletModuleBindings(Guice_servletTest.java:210)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.servletModuleExposesLinkedAndInstanceBindingsWithMatchingRules(Guice_servletTest.java:67)

        Caused by:
        java.lang.ExceptionInInitializerError: Exception java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6b5966e1 [in thread "junit-timeout-thread-3"]
            at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
            at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:203)
            at java.base/java.lang.reflect.Method.setAccessible(Method.java:197)
            at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
            at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
            at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
            at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
            at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
            at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
            at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
            at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
            at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
            at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
            at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
            at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
            at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
            at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
            at com.google.inject.AbstractModule.install(AbstractModule.java:122)
            at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
            at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
            at com.google.inject.spi.Elements.getElements(Elements.java:110)
            at com.google.inject.spi.Elements.getElements(Elements.java:87)
            at com_google_inject_extensions.guice_servlet.Guice_servletTest.duplicateServletMappingsAreReportedByServletModuleElements(Guice_servletTest.java:141)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date

6 tests completed, 2 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 4.0 at 2026-05-18T13:31:06Z with exit code 1 after 5s
❌ ERROR: 'java run' for 4.0 failed with exit code 1.
FAILED[java run][4.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"]
```

## java run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/26036219276/job/76536247387
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
uice-servlet:3.0:4.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 4.0 at 2026-05-18T13:30:55Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean compileTestJava -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 849ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE

> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 1s
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 4.0 at 2026-05-18T13:30:58Z with exit code 0 after 3s
========================================================================================
 com.google.inject.extensions:guice-servlet:3.0:4.0 stage java run
========================================================================================
Starting stage 'java run' for 4.0 at 2026-05-18T13:30:58Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 782ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java uses or overrides a deprecated API.

> Task :compileTestJava
Note: Recompile with -Xlint:deprecation for details.

> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Guice_servletTest > duplicateServletMappingsAreReportedByServletModuleElements() FAILED
    java.lang.ExceptionInInitializerError
        at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
        at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
        at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
        at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
        at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
        at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
        at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
        at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
        at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.spi.Elements.getElements(Elements.java:87)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.duplicateServletMappingsAreReportedByServletModuleElements(Guice_servletTest.java:141)

        Caused by:
        java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @65e61854
            at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
            at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:203)
            at java.base/java.lang.reflect.Method.setAccessible(Method.java:197)
            at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
            at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
            at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
            ... 20 more

Guice_servletTest > servletModuleExposesLinkedAndInstanceBindingsWithMatchingRules() FAILED
    java.lang.NoClassDefFoundError: Could not initialize class com.google.inject.internal.cglib.core.$ReflectUtils
        at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
        at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
        at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
        at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
        at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
        at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
        at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
        at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
        at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.spi.Elements.getElements(Elements.java:87)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.servletModuleBindings(Guice_servletTest.java:210)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.servletModuleExposesLinkedAndInstanceBindingsWithMatchingRules(Guice_servletTest.java:67)

        Caused by:
        java.lang.ExceptionInInitializerError: Exception java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @65e61854 [in thread "junit-timeout-thread-3"]
            at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
            at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:203)
            at java.base/java.lang.reflect.Method.setAccessible(Method.java:197)
            at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
            at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
            at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
            at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
            at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
            at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
            at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
            at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
            at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
            at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
            at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
            at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
            at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
            at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
            at com.google.inject.AbstractModule.install(AbstractModule.java:122)
            at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
            at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
            at com.google.inject.spi.Elements.getElements(Elements.java:110)
            at com.google.inject.spi.Elements.getElements(Elements.java:87)
            at com_google_inject_extensions.guice_servlet.Guice_servletTest.duplicateServletMappingsAreReportedByServletModuleElements(Guice_servletTest.java:141)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

6 tests completed, 2 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 4.0 at 2026-05-18T13:31:02Z with exit code 1 after 4s
❌ ERROR: 'java run' for 4.0 failed with exit code 1.
FAILED[java run][4.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"]
```

## java run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/26036219276/job/76536247477
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
ice-servlet:3.0:4.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 4.0 at 2026-05-18T13:31:25Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.0" ./gradlew clean compileTestJava -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 700ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE

> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 946ms
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 4.0 at 2026-05-18T13:31:27Z with exit code 0 after 2s
========================================================================================
 com.google.inject.extensions:guice-servlet:3.0:4.0 stage java run
========================================================================================
Starting stage 'java run' for 4.0 at 2026-05-18T13:31:27Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 641ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE

> Task :compileTestJava

Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Guice_servletTest > duplicateServletMappingsAreReportedByServletModuleElements() FAILED
    java.lang.ExceptionInInitializerError
        at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
        at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
        at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
        at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
        at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
        at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
        at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
        at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
        at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.spi.Elements.getElements(Elements.java:87)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.duplicateServletMappingsAreReportedByServletModuleElements(Guice_servletTest.java:141)

        Caused by:
        java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @65e61854
            at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
            at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:203)
            at java.base/java.lang.reflect.Method.setAccessible(Method.java:197)
6 tests completed, 2 failed

FAILURE: 
            at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
            at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
            at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
            ... 20 more

Guice_servletTest > servletModuleExposesLinkedAndInstanceBindingsWithMatchingRules() FAILED
    java.lang.NoClassDefFoundError: Could not initialize class com.google.inject.internal.cglib.core.$ReflectUtils
        at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
        at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
        at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
        at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
        at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
        at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
        at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
        at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
        at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
        at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.spi.Elements.getElements(Elements.java:87)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.servletModuleBindings(Guice_servletTest.java:210)
        at com_google_inject_extensions.guice_servlet.Guice_servletTest.servletModuleExposesLinkedAndInstanceBindingsWithMatchingRules(Guice_servletTest.java:67)

        Caused by:
        java.lang.ExceptionInInitializerError: Exception java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @65e61854 [in thread "junit-timeout-thread-3"]
            at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
            at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:203)
            at java.base/java.lang.reflect.Method.setAccessible(Method.java:197)
            at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
            at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
            at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
            at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
            at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
            at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
            at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
            at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
            at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
            at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
            at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
            at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
            at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
            at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
            at com.google.inject.AbstractModule.install(AbstractModule.java:122)
            at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
            at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
            at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
            at com.google.inject.spi.Elements.getElements(Elements.java:110)
            at com.google.inject.spi.Elements.getElements(Elements.java:87)
            at com_google_inject_extensions.guice_servlet.Guice_servletTest.duplicateServletMappingsAreReportedByServletModuleElements(Guice_servletTest.java:141)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date
Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.google.inject.extensions/guice-servlet/3.0/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 3s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 4.0 at 2026-05-18T13:31:31Z with exit code 1 after 4s
❌ ERROR: 'java run' for 4.0 failed with exit code 1.
FAILED[java run][4.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.0" ./gradlew clean javaTest -Pcoordinates="com.google.inject.extensions:guice-servlet:3.0"]
```

--- a/metadata/com.google.inject.extensions/guice-servlet/4.0/reachability-metadata.json
+++ b/metadata/com.google.inject.extensions/guice-servlet/4.0/reachability-metadata.json
@@ -1,0 +1,199 @@
+{
+  "reflection": [
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.util.$FinalizableReferenceQueue"
+      },
+      "type": "com.google.inject.internal.util.$Finalizer",
+      "methods": [
+        {
+          "name": "startFinalizer",
+          "parameterTypes": [
+            "java.lang.Class",
+            "java.lang.Object"
+          ]
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.Annotations"
+      },
+      "type": "com.google.inject.name.Named"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.Annotations"
+      },
+      "type": "com.google.inject.servlet.RequestParameters"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.Annotations"
+      },
+      "type": "com.google.inject.internal.UniqueAnnotations$Internal"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.servlet.InternalServletModule"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.Binding"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.MembersInjector"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.Provider"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.TypeLiteral"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.matcher.Matcher"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.spi.Message"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.spi.ProvisionListener"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.spi.TypeConverter"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.spi.TypeListener"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.internal.ProviderMethod"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.internal.ProviderMethod$ReflectionProviderMethod"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.spi.BindingTargetVisitor"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.spi.Dependency"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.spi.HasDependencies"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.spi.ProvidesMethodBinding"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.spi.ProvidesMethodTargetVisitor"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.spi.ProviderInstanceBinding"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.spi.ProviderWithExtensionVisitor"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.servlet.FilterDefinition"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.spi.InjectionPoint"
+      },
+      "type": "com.google.inject.servlet.ServletDefinition"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.util.Modules$CombinedModule"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.util.Modules$EmptyModule"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.util.Modules$ModuleWriter"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.util.Modules$OverrideModule"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.util.Modules$OverriddenModuleBuilder"
+    },
+    {
+      "condition": {
+        "typeReached": "com.google.inject.internal.ProviderMethodsModule"
+      },
+      "type": "com.google.inject.util.Modules$RealOverriddenModuleBuilder"
+    }
+  ]
+}

--- a/metadata/com.google.inject.extensions/guice-servlet/index.json
+++ b/metadata/com.google.inject.extensions/guice-servlet/index.json
@@ -2,12 +2,16 @@
   {
     "latest" : true,
     "metadata-version" : "4.0",
+    "source-code-url" : "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-sources.jar",
+    "repository-url" : "https://github.com/google/guice",
+    "test-code-url" : "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-tests.jar",
+    "documentation-url" : "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-javadoc.jar",
+    "description" : "Guice Servlet extends Google Guice for Java servlet applications, adding request and session scopes plus servlet and filter binding support. It integrates Guice-managed dependency injection into servlet containers through APIs such as ServletModule, GuiceFilter, and GuiceServletContextListener.",
     "tested-versions" : [
       "4.0"
     ],
     "allowed-packages" : [
-      "com.google.inject.servlet",
-      "com.google.inject.internal.util"
+      "com.google.inject"
     ]
   },
   {
@@ -21,8 +25,7 @@
       "3.0"
     ],
     "allowed-packages" : [
-      "com.google.inject.servlet",
-      "com.google.inject.internal.util"
+      "com.google.inject"
     ]
   }
 ]

--- a/metadata/com.google.inject.extensions/guice-servlet/index.json
+++ b/metadata/com.google.inject.extensions/guice-servlet/index.json
@@ -1,16 +1,26 @@
 [
   {
-    "latest": true,
-    "metadata-version": "3.0",
-    "source-code-url": "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-sources.jar",
-    "repository-url": "https://github.com/google/guice",
-    "test-code-url": "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-tests.jar",
-    "documentation-url": "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-javadoc.jar",
-    "description": "Guice Servlet extends Google Guice for Java servlet applications, adding request and session scopes plus servlet and filter binding support. It integrates Guice-managed dependency injection into servlet containers through APIs such as ServletModule, GuiceFilter, and GuiceServletContextListener.",
-    "tested-versions": [
+    "latest" : true,
+    "metadata-version" : "4.0",
+    "tested-versions" : [
+      "4.0"
+    ],
+    "allowed-packages" : [
+      "com.google.inject.servlet",
+      "com.google.inject.internal.util"
+    ]
+  },
+  {
+    "metadata-version" : "3.0",
+    "source-code-url" : "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-sources.jar",
+    "repository-url" : "https://github.com/google/guice",
+    "test-code-url" : "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-tests.jar",
+    "documentation-url" : "https://repo1.maven.org/maven2/com/google/inject/extensions/guice-servlet/$version$/guice-servlet-$version$-javadoc.jar",
+    "description" : "Guice Servlet extends Google Guice for Java servlet applications, adding request and session scopes plus servlet and filter binding support. It integrates Guice-managed dependency injection into servlet containers through APIs such as ServletModule, GuiceFilter, and GuiceServletContextListener.",
+    "tested-versions" : [
       "3.0"
     ],
-    "allowed-packages": [
+    "allowed-packages" : [
       "com.google.inject.servlet",
       "com.google.inject.internal.util"
     ]

--- a/stats/com.google.inject.extensions/guice-servlet/4.0/execution-metrics.json
+++ b/stats/com.google.inject.extensions/guice-servlet/4.0/execution-metrics.json
@@ -1,0 +1,87 @@
+{
+  "fix_java_run_fail:2026-05-19": {
+    "timestamp": "2026-05-19T00:40:29.074080Z",
+    "library": "com.google.inject.extensions:guice-servlet:4.0",
+    "starting_commit": "7fd825e4b3270ddf7fc513e7da12e9e98ad25cc9",
+    "ending_commit": "65f6a210ef75c3ecbc57f73aea9b9935e189f6b1",
+    "previous_library": "com.google.inject.extensions:guice-servlet:3.0",
+    "strategy_name": "java_run_iterative_with_coverage_sources_pi_gpt-5.5",
+    "agent": "pi",
+    "model": "oca/gpt-5.5",
+    "status": "success",
+    "stats": {
+      "version": "4.0",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 1486,
+          "missed": 1767,
+          "ratio": 0.456809,
+          "total": 3253
+        },
+        "line": {
+          "covered": 288,
+          "missed": 371,
+          "ratio": 0.437026,
+          "total": 659
+        },
+        "method": {
+          "covered": 100,
+          "missed": 122,
+          "ratio": 0.45045,
+          "total": 222
+        }
+      }
+    },
+    "previous_library_stats": {
+      "version": "3.0",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 1353,
+          "missed": 1276,
+          "ratio": 0.514644,
+          "total": 2629
+        },
+        "line": {
+          "covered": 272,
+          "missed": 262,
+          "ratio": 0.509363,
+          "total": 534
+        },
+        "method": {
+          "covered": 89,
+          "missed": 95,
+          "ratio": 0.483696,
+          "total": 184
+        }
+      }
+    },
+    "metrics": {
+      "input_tokens_used": 101325,
+      "cached_input_tokens_used": 401408,
+      "output_tokens_used": 3371,
+      "iterations": 1,
+      "cost_usd": 0.3018,
+      "tested_library_loc": 288,
+      "metadata_entries": 32,
+      "previous_library_metadata_entries": 2,
+      "code_coverage_percent": 43.7,
+      "previous_library_coverage_percent": 50.94
+    },
+    "artifacts": {
+      "test_file": "tests/src/com.google.inject.extensions/guice-servlet/4.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java",
+      "metadata_file": "metadata/com.google.inject.extensions/guice-servlet/4.0/reachability-metadata.json"
+    }
+  }
+}

--- a/stats/com.google.inject.extensions/guice-servlet/4.0/stats.json
+++ b/stats/com.google.inject.extensions/guice-servlet/4.0/stats.json
@@ -1,0 +1,31 @@
+{
+  "versions" : [ {
+    "version" : "4.0",
+    "dynamicAccess" : {
+      "breakdown" : { },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 0,
+      "totalCalls" : 0
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 1486,
+        "missed" : 1767,
+        "ratio" : 0.456809,
+        "total" : 3253
+      },
+      "line" : {
+        "covered" : 288,
+        "missed" : 371,
+        "ratio" : 0.437026,
+        "total" : 659
+      },
+      "method" : {
+        "covered" : 100,
+        "missed" : 122,
+        "ratio" : 0.45045,
+        "total" : 222
+      }
+    }
+  } ]
+}

--- a/tests/src/com.google.inject.extensions/guice-servlet/4.0/.gitignore
+++ b/tests/src/com.google.inject.extensions/guice-servlet/4.0/.gitignore
@@ -1,0 +1,4 @@
+gradlew.bat
+gradlew
+gradle/
+build/

--- a/tests/src/com.google.inject.extensions/guice-servlet/4.0/build.gradle
+++ b/tests/src/com.google.inject.extensions/guice-servlet/4.0/build.gradle
@@ -1,0 +1,28 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "com.google.inject.extensions:guice-servlet:$libraryVersion"
+    testImplementation 'javax.servlet:servlet-api:2.5'
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+    }
+}

--- a/tests/src/com.google.inject.extensions/guice-servlet/4.0/build.gradle
+++ b/tests/src/com.google.inject.extensions/guice-servlet/4.0/build.gradle
@@ -11,7 +11,10 @@ plugins {
 String libraryVersion = tck.testedLibraryVersion.get()
 
 dependencies {
-    testImplementation "com.google.inject.extensions:guice-servlet:$libraryVersion"
+    testImplementation("com.google.inject.extensions:guice-servlet:$libraryVersion") {
+        exclude group: 'com.google.inject', module: 'guice'
+    }
+    testImplementation "com.google.inject:guice:$libraryVersion:no_aop"
     testImplementation 'javax.servlet:servlet-api:2.5'
     testImplementation 'org.assertj:assertj-core:3.22.0'
 }

--- a/tests/src/com.google.inject.extensions/guice-servlet/4.0/gradle.properties
+++ b/tests/src/com.google.inject.extensions/guice-servlet/4.0/gradle.properties
@@ -1,0 +1,6 @@
+# Optional explicit tested library coordinates (format: group:artifact:version).
+# If omitted, fallback resolution in org.graalvm.internal.tck.gradle still uses
+# environment variable GVM_TCK_LC and then this property.
+library.coordinates = com.google.inject.extensions:guice-servlet:4.0
+library.version = 4.0
+metadata.dir = com.google.inject.extensions/guice-servlet/4.0/

--- a/tests/src/com.google.inject.extensions/guice-servlet/4.0/settings.gradle
+++ b/tests/src/com.google.inject.extensions/guice-servlet/4.0/settings.gradle
@@ -1,0 +1,20 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+    repositories { mavenLocal(); gradlePluginPortal() }
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'com.google.inject.extensions.guice-servlet_tests'

--- a/tests/src/com.google.inject.extensions/guice-servlet/4.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java
+++ b/tests/src/com.google.inject.extensions/guice-servlet/4.0/src/test/java/com_google_inject_extensions/guice_servlet/Guice_servletTest.java
@@ -1,0 +1,947 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package com_google_inject_extensions.guice_servlet;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import com.google.inject.Binding;
+import com.google.inject.Key;
+import com.google.inject.OutOfScopeException;
+import com.google.inject.Provider;
+import com.google.inject.Singleton;
+import com.google.inject.name.Names;
+import com.google.inject.servlet.GuiceFilter;
+import com.google.inject.servlet.InstanceFilterBinding;
+import com.google.inject.servlet.InstanceServletBinding;
+import com.google.inject.servlet.LinkedFilterBinding;
+import com.google.inject.servlet.LinkedServletBinding;
+import com.google.inject.servlet.ServletModule;
+import com.google.inject.servlet.ServletModuleBinding;
+import com.google.inject.servlet.ServletModuleTargetVisitor;
+import com.google.inject.servlet.ServletScopes;
+import com.google.inject.servlet.UriPatternType;
+import com.google.inject.spi.DefaultBindingTargetVisitor;
+import com.google.inject.spi.Element;
+import com.google.inject.spi.Elements;
+import com.google.inject.spi.Message;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.security.Principal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.ServletInputStream;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import javax.servlet.http.HttpSessionContext;
+import org.junit.jupiter.api.Test;
+
+public class Guice_servletTest {
+    private static final Key<String> REQUEST_USER_KEY = Key.get(String.class, Names.named("requestUser"));
+
+    @Test
+    void servletModuleExposesLinkedAndInstanceBindingsWithMatchingRules() {
+        List<ServletModuleBinding> bindings = servletModuleBindings(new RoutingModule());
+
+        assertThat(bindings).hasSize(7);
+
+        LinkedFilterBinding apiFilter = binding(bindings, LinkedFilterBinding.class, "/api/*");
+        assertThat(apiFilter.getUriPatternType()).isEqualTo(UriPatternType.SERVLET);
+        assertThat(apiFilter.getLinkedKey()).isEqualTo(Key.get(ApiFilter.class));
+        assertThat(apiFilter.getInitParams()).containsEntry("filter", "api");
+        assertThat(apiFilter.matchesUri("/api/customers")).isTrue();
+        assertThat(apiFilter.matchesUri("/admin/42")).isFalse();
+
+        LinkedFilterBinding htmlFilter = binding(bindings, LinkedFilterBinding.class, "*.html");
+        assertThat(htmlFilter.matchesUri("/index.html")).isTrue();
+        assertThat(htmlFilter.matchesUri("/index.json")).isFalse();
+
+        InstanceFilterBinding adminFilter = binding(bindings, InstanceFilterBinding.class, "^/admin/[0-9]+$");
+        assertThat(adminFilter.getUriPatternType()).isEqualTo(UriPatternType.REGEX);
+        assertThat(adminFilter.getFilterInstance()).isSameAs(RoutingModule.ADMIN_FILTER);
+        assertThat(adminFilter.getInitParams()).containsEntry("filter", "admin");
+        assertThat(adminFilter.matchesUri("/admin/42")).isTrue();
+        assertThat(adminFilter.matchesUri("/admin/users")).isFalse();
+
+        LinkedServletBinding healthServlet = binding(bindings, LinkedServletBinding.class, "/health");
+        assertThat(healthServlet.getLinkedKey()).isEqualTo(Key.get(HealthServlet.class));
+        assertThat(healthServlet.getInitParams()).containsEntry("servlet", "health");
+        assertThat(healthServlet.matchesUri("/health")).isTrue();
+        assertThat(healthServlet.matchesUri("/health/check")).isFalse();
+
+        LinkedServletBinding staticServlet = binding(bindings, LinkedServletBinding.class, "/static/*");
+        assertThat(staticServlet.matchesUri("/static/app.js")).isTrue();
+        assertThat(staticServlet.matchesUri("/assets/app.js")).isFalse();
+
+        InstanceServletBinding assetServlet = binding(bindings, InstanceServletBinding.class, "^/assets/.+");
+        assertThat(assetServlet.getUriPatternType()).isEqualTo(UriPatternType.REGEX);
+        assertThat(assetServlet.getServletInstance()).isSameAs(RoutingModule.ASSET_SERVLET);
+        assertThat(assetServlet.getInitParams()).containsEntry("servlet", "asset");
+        assertThat(assetServlet.matchesUri("/assets/app.css")).isTrue();
+        assertThat(assetServlet.matchesUri("/static/app.css")).isFalse();
+
+        LinkedServletBinding namedServlet = binding(bindings, LinkedServletBinding.class, "/named/*");
+        assertThat(namedServlet.getLinkedKey()).isEqualTo(Key.get(NamedServlet.class, Names.named("namedServlet")));
+    }
+
+    @Test
+    void requestScopeCachesValuesAndUsesSeededRequestData() throws Exception {
+        Provider<RequestScopedValue> requestScopedProvider = ServletScopes.REQUEST.scope(
+                Key.get(RequestScopedValue.class), new RequestScopedValueProvider());
+        Provider<String> userProvider = ServletScopes.REQUEST.scope(REQUEST_USER_KEY, new DefaultRequestUserProvider());
+
+        assertThatThrownBy(requestScopedProvider::get).isInstanceOf(OutOfScopeException.class);
+
+        ScopedRequestResult first = executeInRequestScope(requestScopedProvider, userProvider, "alice");
+        ScopedRequestResult second = executeInRequestScope(requestScopedProvider, userProvider, "bob");
+
+        assertThat(first.firstValue).isSameAs(first.secondValue);
+        assertThat(first.user).isEqualTo("alice");
+        assertThat(second.firstValue).isSameAs(second.secondValue);
+        assertThat(second.user).isEqualTo("bob");
+        assertThat(first.firstValue).isNotSameAs(second.firstValue);
+    }
+
+    @Test
+    void requestScopeRejectsSeedValuesWithTheWrongType() {
+        Map<Key<?>, Object> seedMap = new LinkedHashMap<>();
+        seedMap.put(REQUEST_USER_KEY, 42);
+
+        assertThatThrownBy(() -> ServletScopes.scopeRequest(() -> null, seedMap))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("not compatible")
+                .hasMessageContaining(REQUEST_USER_KEY.toString());
+    }
+
+    @Test
+    void duplicateServletMappingsAreReportedByServletModuleElements() {
+        List<Message> messages = Elements.getElements(new DuplicateServletMappingModule()).stream()
+                .filter(Message.class::isInstance)
+                .map(Message.class::cast)
+                .toList();
+
+        assertThat(messages.stream().map(Message::getMessage).toList())
+                .anySatisfy(message -> assertThat(message)
+                        .contains("More than one servlet was mapped")
+                        .contains("/duplicate"));
+    }
+
+    @Test
+    void sessionScopeCachesValuesAcrossRequestsUsingTheSameHttpSession() throws Exception {
+        Provider<SessionScopedValue> sessionScopedProvider = ServletScopes.SESSION.scope(
+                Key.get(SessionScopedValue.class), new SessionScopedValueProvider());
+        TestHttpSession session = new TestHttpSession();
+        TestHttpSession otherSession = new TestHttpSession();
+
+        ScopedSessionResult first = executeWithSession(session, () -> new ScopedSessionResult(
+                sessionScopedProvider.get(), sessionScopedProvider.get()));
+        ScopedSessionResult second = executeWithSession(session, () -> new ScopedSessionResult(
+                sessionScopedProvider.get(), sessionScopedProvider.get()));
+        ScopedSessionResult third = executeWithSession(otherSession, () -> new ScopedSessionResult(
+                sessionScopedProvider.get(), sessionScopedProvider.get()));
+
+        assertThat(first.firstValue).isSameAs(first.secondValue);
+        assertThat(second.firstValue).isSameAs(first.firstValue);
+        assertThat(third.firstValue).isSameAs(third.secondValue);
+        assertThat(third.firstValue).isNotSameAs(first.firstValue);
+    }
+
+    @Test
+    void continueRequestCreatesCallableThatUsesRequestScopeAfterServletFilterReturns() throws Exception {
+        Provider<RequestScopedValue> requestScopedProvider = ServletScopes.REQUEST.scope(
+                Key.get(RequestScopedValue.class), new RequestScopedValueProvider());
+        Provider<String> userProvider = ServletScopes.REQUEST.scope(REQUEST_USER_KEY, new DefaultRequestUserProvider());
+        Provider<SessionScopedValue> sessionScopedProvider = ServletScopes.SESSION.scope(
+                Key.get(SessionScopedValue.class), new SessionScopedValueProvider());
+        ContinuedRequestCapture capture = new ContinuedRequestCapture(
+                requestScopedProvider, userProvider, sessionScopedProvider);
+
+        new GuiceFilter().doFilter(
+                new TestHttpServletRequest(new TestHttpSession()), new TestHttpServletResponse(), capture);
+
+        ContinuedRequestResult result = capture.continuedRequest.call();
+
+        assertThat(result.firstValue).isSameAs(result.secondValue);
+        assertThat(result.user).isEqualTo(ContinuedRequestCapture.CONTINUED_USER);
+        assertThatThrownBy(() -> capture.continuedSessionAccess.call())
+                .isInstanceOf(OutOfScopeException.class)
+                .hasMessageContaining("Cannot access the session");
+    }
+
+    private static ScopedRequestResult executeInRequestScope(
+            Provider<RequestScopedValue> requestScopedProvider,
+            Provider<String> userProvider,
+            String user) throws Exception {
+        Map<Key<?>, Object> seedMap = new LinkedHashMap<>();
+        seedMap.put(REQUEST_USER_KEY, user);
+
+        return ServletScopes.scopeRequest(() -> new ScopedRequestResult(
+                requestScopedProvider.get(),
+                requestScopedProvider.get(),
+                userProvider.get()), seedMap).call();
+    }
+
+    private static List<ServletModuleBinding> servletModuleBindings(ServletModule module) {
+        List<ServletModuleBinding> bindings = new ArrayList<>();
+        BindingCollector collector = new BindingCollector();
+        for (Element element : Elements.getElements(module)) {
+            if (element instanceof Binding) {
+                ServletModuleBinding servletBinding = ((Binding<?>) element).acceptTargetVisitor(collector);
+                if (servletBinding != null) {
+                    bindings.add(servletBinding);
+                }
+            }
+        }
+        return bindings;
+    }
+
+    private static <T extends ServletModuleBinding> T binding(
+            List<ServletModuleBinding> bindings, Class<T> bindingType, String pattern) {
+        return bindings.stream()
+                .filter(bindingType::isInstance)
+                .map(bindingType::cast)
+                .filter(binding -> pattern.equals(binding.getPattern()))
+                .findFirst()
+                .orElseThrow(() -> new AssertionError("No " + bindingType.getSimpleName() + " for " + pattern));
+    }
+
+    private static final class BindingCollector extends DefaultBindingTargetVisitor<Object, ServletModuleBinding>
+            implements ServletModuleTargetVisitor<Object, ServletModuleBinding> {
+        @Override
+        public ServletModuleBinding visit(LinkedFilterBinding binding) {
+            return binding;
+        }
+
+        @Override
+        public ServletModuleBinding visit(InstanceFilterBinding binding) {
+            return binding;
+        }
+
+        @Override
+        public ServletModuleBinding visit(LinkedServletBinding binding) {
+            return binding;
+        }
+
+        @Override
+        public ServletModuleBinding visit(InstanceServletBinding binding) {
+            return binding;
+        }
+    }
+
+    private static final class RoutingModule extends ServletModule {
+        private static final Filter ADMIN_FILTER = new AdminFilter();
+        private static final HttpServlet ASSET_SERVLET = new AssetServlet();
+
+        @Override
+        protected void configureServlets() {
+            Map<String, String> apiFilterParams = new LinkedHashMap<>();
+            apiFilterParams.put("filter", "api");
+            filter("/api/*", "*.html").through(ApiFilter.class, apiFilterParams);
+
+            Map<String, String> adminFilterParams = new LinkedHashMap<>();
+            adminFilterParams.put("filter", "admin");
+            filterRegex("^/admin/[0-9]+$").through(ADMIN_FILTER, adminFilterParams);
+
+            Map<String, String> healthServletParams = new LinkedHashMap<>();
+            healthServletParams.put("servlet", "health");
+            serve("/health", "/static/*").with(HealthServlet.class, healthServletParams);
+
+            Map<String, String> assetServletParams = new LinkedHashMap<>();
+            assetServletParams.put("servlet", "asset");
+            serveRegex("^/assets/.+").with(ASSET_SERVLET, assetServletParams);
+
+            bind(NamedServlet.class)
+                    .annotatedWith(Names.named("namedServlet"))
+                    .to(NamedServlet.class)
+                    .in(Singleton.class);
+            serve("/named/*").with(Key.get(NamedServlet.class, Names.named("namedServlet")));
+        }
+    }
+
+    private static final class DuplicateServletMappingModule extends ServletModule {
+        @Override
+        protected void configureServlets() {
+            serve("/duplicate").with(HealthServlet.class);
+            serve("/duplicate").with(AssetServlet.class);
+        }
+    }
+
+    private static <T> T executeWithSession(HttpSession session, ValueSupplier<T> supplier)
+            throws IOException, ServletException {
+        ValueCapturingFilterChain<T> chain = new ValueCapturingFilterChain<>(supplier);
+        new GuiceFilter().doFilter(new TestHttpServletRequest(session), new TestHttpServletResponse(), chain);
+        return chain.value;
+    }
+
+    private interface ValueSupplier<T> {
+        T get();
+    }
+
+    private static final class ValueCapturingFilterChain<T> implements FilterChain {
+        private final ValueSupplier<T> supplier;
+        private T value;
+
+        private ValueCapturingFilterChain(ValueSupplier<T> supplier) {
+            this.supplier = supplier;
+        }
+
+        @Override
+        public void doFilter(ServletRequest request, ServletResponse response) {
+            value = supplier.get();
+        }
+    }
+
+    private static final class ContinuedRequestCapture implements FilterChain {
+        private static final String CONTINUED_USER = "continued-user";
+
+        private final Provider<RequestScopedValue> requestScopedProvider;
+        private final Provider<String> userProvider;
+        private final Provider<SessionScopedValue> sessionScopedProvider;
+        private Callable<ContinuedRequestResult> continuedRequest;
+        private Callable<SessionScopedValue> continuedSessionAccess;
+
+        private ContinuedRequestCapture(
+                Provider<RequestScopedValue> requestScopedProvider,
+                Provider<String> userProvider,
+                Provider<SessionScopedValue> sessionScopedProvider) {
+            this.requestScopedProvider = requestScopedProvider;
+            this.userProvider = userProvider;
+            this.sessionScopedProvider = sessionScopedProvider;
+        }
+
+        @Override
+        public void doFilter(ServletRequest request, ServletResponse response) {
+            Map<Key<?>, Object> seedMap = new LinkedHashMap<>();
+            seedMap.put(REQUEST_USER_KEY, CONTINUED_USER);
+            continuedRequest = ServletScopes.continueRequest(() -> new ContinuedRequestResult(
+                    requestScopedProvider.get(), requestScopedProvider.get(), userProvider.get()), seedMap);
+            continuedSessionAccess = ServletScopes.continueRequest(sessionScopedProvider::get, Collections.emptyMap());
+        }
+    }
+
+    private static final class ContinuedRequestResult {
+        private final RequestScopedValue firstValue;
+        private final RequestScopedValue secondValue;
+        private final String user;
+
+        private ContinuedRequestResult(RequestScopedValue firstValue, RequestScopedValue secondValue, String user) {
+            this.firstValue = firstValue;
+            this.secondValue = secondValue;
+            this.user = user;
+        }
+    }
+
+    private static final class ScopedSessionResult {
+        private final SessionScopedValue firstValue;
+        private final SessionScopedValue secondValue;
+
+        private ScopedSessionResult(SessionScopedValue firstValue, SessionScopedValue secondValue) {
+            this.firstValue = firstValue;
+            this.secondValue = secondValue;
+        }
+    }
+
+    private static final class TestHttpSession implements HttpSession {
+        private final Map<String, Object> attributes = new LinkedHashMap<>();
+
+        @Override
+        public long getCreationTime() {
+            return 0L;
+        }
+
+        @Override
+        public String getId() {
+            return "test-session";
+        }
+
+        @Override
+        public long getLastAccessedTime() {
+            return 0L;
+        }
+
+        @Override
+        public ServletContext getServletContext() {
+            return null;
+        }
+
+        @Override
+        public void setMaxInactiveInterval(int interval) {
+        }
+
+        @Override
+        public int getMaxInactiveInterval() {
+            return 0;
+        }
+
+        @Override
+        @SuppressWarnings("deprecation")
+        public HttpSessionContext getSessionContext() {
+            return null;
+        }
+
+        @Override
+        public Object getAttribute(String name) {
+            return attributes.get(name);
+        }
+
+        @Override
+        public Object getValue(String name) {
+            return getAttribute(name);
+        }
+
+        @Override
+        public Enumeration<String> getAttributeNames() {
+            return Collections.enumeration(attributes.keySet());
+        }
+
+        @Override
+        public String[] getValueNames() {
+            return attributes.keySet().toArray(new String[0]);
+        }
+
+        @Override
+        public void setAttribute(String name, Object value) {
+            attributes.put(name, value);
+        }
+
+        @Override
+        public void putValue(String name, Object value) {
+            setAttribute(name, value);
+        }
+
+        @Override
+        public void removeAttribute(String name) {
+            attributes.remove(name);
+        }
+
+        @Override
+        public void removeValue(String name) {
+            removeAttribute(name);
+        }
+
+        @Override
+        public void invalidate() {
+            attributes.clear();
+        }
+
+        @Override
+        public boolean isNew() {
+            return false;
+        }
+    }
+
+    private static final class TestHttpServletRequest implements HttpServletRequest {
+        private final HttpSession session;
+        private final Map<String, Object> attributes = new LinkedHashMap<>();
+
+        private TestHttpServletRequest(HttpSession session) {
+            this.session = session;
+        }
+
+        @Override
+        public String getAuthType() {
+            return null;
+        }
+
+        @Override
+        public Cookie[] getCookies() {
+            return new Cookie[0];
+        }
+
+        @Override
+        public long getDateHeader(String name) {
+            return -1L;
+        }
+
+        @Override
+        public String getHeader(String name) {
+            return null;
+        }
+
+        @Override
+        public Enumeration<String> getHeaders(String name) {
+            return Collections.emptyEnumeration();
+        }
+
+        @Override
+        public Enumeration<String> getHeaderNames() {
+            return Collections.emptyEnumeration();
+        }
+
+        @Override
+        public int getIntHeader(String name) {
+            return -1;
+        }
+
+        @Override
+        public String getMethod() {
+            return "GET";
+        }
+
+        @Override
+        public String getPathInfo() {
+            return null;
+        }
+
+        @Override
+        public String getPathTranslated() {
+            return null;
+        }
+
+        @Override
+        public String getContextPath() {
+            return "";
+        }
+
+        @Override
+        public String getQueryString() {
+            return null;
+        }
+
+        @Override
+        public String getRemoteUser() {
+            return null;
+        }
+
+        @Override
+        public boolean isUserInRole(String role) {
+            return false;
+        }
+
+        @Override
+        public Principal getUserPrincipal() {
+            return null;
+        }
+
+        @Override
+        public String getRequestedSessionId() {
+            return session.getId();
+        }
+
+        @Override
+        public String getRequestURI() {
+            return "/session";
+        }
+
+        @Override
+        public StringBuffer getRequestURL() {
+            return new java.lang.StringBuffer("http://localhost/session");
+        }
+
+        @Override
+        public String getServletPath() {
+            return "/session";
+        }
+
+        @Override
+        public HttpSession getSession(boolean create) {
+            return session;
+        }
+
+        @Override
+        public HttpSession getSession() {
+            return session;
+        }
+
+        @Override
+        public boolean isRequestedSessionIdValid() {
+            return true;
+        }
+
+        @Override
+        public boolean isRequestedSessionIdFromCookie() {
+            return false;
+        }
+
+        @Override
+        public boolean isRequestedSessionIdFromURL() {
+            return false;
+        }
+
+        @Override
+        @SuppressWarnings("deprecation")
+        public boolean isRequestedSessionIdFromUrl() {
+            return false;
+        }
+
+        @Override
+        public Object getAttribute(String name) {
+            return attributes.get(name);
+        }
+
+        @Override
+        public Enumeration<String> getAttributeNames() {
+            return Collections.enumeration(attributes.keySet());
+        }
+
+        @Override
+        public String getCharacterEncoding() {
+            return "UTF-8";
+        }
+
+        @Override
+        public void setCharacterEncoding(String encoding) {
+        }
+
+        @Override
+        public int getContentLength() {
+            return 0;
+        }
+
+        @Override
+        public String getContentType() {
+            return null;
+        }
+
+        @Override
+        public ServletInputStream getInputStream() {
+            return null;
+        }
+
+        @Override
+        public String getParameter(String name) {
+            return null;
+        }
+
+        @Override
+        public Enumeration<String> getParameterNames() {
+            return Collections.emptyEnumeration();
+        }
+
+        @Override
+        public String[] getParameterValues(String name) {
+            return null;
+        }
+
+        @Override
+        public Map<String, String[]> getParameterMap() {
+            return Collections.emptyMap();
+        }
+
+        @Override
+        public String getProtocol() {
+            return "HTTP/1.1";
+        }
+
+        @Override
+        public String getScheme() {
+            return "http";
+        }
+
+        @Override
+        public String getServerName() {
+            return "localhost";
+        }
+
+        @Override
+        public int getServerPort() {
+            return 80;
+        }
+
+        @Override
+        public BufferedReader getReader() {
+            return null;
+        }
+
+        @Override
+        public String getRemoteAddr() {
+            return "127.0.0.1";
+        }
+
+        @Override
+        public String getRemoteHost() {
+            return "localhost";
+        }
+
+        @Override
+        public void setAttribute(String name, Object object) {
+            attributes.put(name, object);
+        }
+
+        @Override
+        public void removeAttribute(String name) {
+            attributes.remove(name);
+        }
+
+        @Override
+        public Locale getLocale() {
+            return Locale.ROOT;
+        }
+
+        @Override
+        public Enumeration<Locale> getLocales() {
+            return Collections.enumeration(List.of(Locale.ROOT));
+        }
+
+        @Override
+        public boolean isSecure() {
+            return false;
+        }
+
+        @Override
+        public RequestDispatcher getRequestDispatcher(String path) {
+            return null;
+        }
+
+        @Override
+        @SuppressWarnings("deprecation")
+        public String getRealPath(String path) {
+            return null;
+        }
+
+        @Override
+        public int getRemotePort() {
+            return 0;
+        }
+
+        @Override
+        public String getLocalName() {
+            return "localhost";
+        }
+
+        @Override
+        public String getLocalAddr() {
+            return "127.0.0.1";
+        }
+
+        @Override
+        public int getLocalPort() {
+            return 80;
+        }
+    }
+
+    private static final class TestHttpServletResponse implements HttpServletResponse {
+        @Override
+        public void addCookie(Cookie cookie) {
+        }
+
+        @Override
+        public boolean containsHeader(String name) {
+            return false;
+        }
+
+        @Override
+        public String encodeURL(String url) {
+            return url;
+        }
+
+        @Override
+        public String encodeRedirectURL(String url) {
+            return url;
+        }
+
+        @Override
+        @SuppressWarnings("deprecation")
+        public String encodeUrl(String url) {
+            return url;
+        }
+
+        @Override
+        @SuppressWarnings("deprecation")
+        public String encodeRedirectUrl(String url) {
+            return url;
+        }
+
+        @Override
+        public void sendError(int statusCode, String message) {
+        }
+
+        @Override
+        public void sendError(int statusCode) {
+        }
+
+        @Override
+        public void sendRedirect(String location) {
+        }
+
+        @Override
+        public void setDateHeader(String name, long date) {
+        }
+
+        @Override
+        public void addDateHeader(String name, long date) {
+        }
+
+        @Override
+        public void setHeader(String name, String value) {
+        }
+
+        @Override
+        public void addHeader(String name, String value) {
+        }
+
+        @Override
+        public void setIntHeader(String name, int value) {
+        }
+
+        @Override
+        public void addIntHeader(String name, int value) {
+        }
+
+        @Override
+        public void setStatus(int statusCode) {
+        }
+
+        @Override
+        @SuppressWarnings("deprecation")
+        public void setStatus(int statusCode, String message) {
+        }
+
+        @Override
+        public String getCharacterEncoding() {
+            return "UTF-8";
+        }
+
+        @Override
+        public String getContentType() {
+            return null;
+        }
+
+        @Override
+        public ServletOutputStream getOutputStream() {
+            return null;
+        }
+
+        @Override
+        public PrintWriter getWriter() {
+            return null;
+        }
+
+        @Override
+        public void setCharacterEncoding(String encoding) {
+        }
+
+        @Override
+        public void setContentLength(int length) {
+        }
+
+        @Override
+        public void setContentType(String type) {
+        }
+
+        @Override
+        public void setBufferSize(int size) {
+        }
+
+        @Override
+        public int getBufferSize() {
+            return 0;
+        }
+
+        @Override
+        public void flushBuffer() {
+        }
+
+        @Override
+        public void resetBuffer() {
+        }
+
+        @Override
+        public boolean isCommitted() {
+            return false;
+        }
+
+        @Override
+        public void reset() {
+        }
+
+        @Override
+        public void setLocale(Locale locale) {
+        }
+
+        @Override
+        public Locale getLocale() {
+            return Locale.ROOT;
+        }
+    }
+
+    private static final class ScopedRequestResult {
+        private final RequestScopedValue firstValue;
+        private final RequestScopedValue secondValue;
+        private final String user;
+
+        private ScopedRequestResult(RequestScopedValue firstValue, RequestScopedValue secondValue, String user) {
+            this.firstValue = firstValue;
+            this.secondValue = secondValue;
+            this.user = user;
+        }
+    }
+
+    private static final class DefaultRequestUserProvider implements Provider<String> {
+        @Override
+        public String get() {
+            return "unseeded-user";
+        }
+    }
+
+    private static final class RequestScopedValueProvider implements Provider<RequestScopedValue> {
+        @Override
+        public RequestScopedValue get() {
+            return new RequestScopedValue();
+        }
+    }
+
+    private static final class SessionScopedValueProvider implements Provider<SessionScopedValue> {
+        @Override
+        public SessionScopedValue get() {
+            return new SessionScopedValue();
+        }
+    }
+
+    private static final class RequestScopedValue {
+    }
+
+    private static final class SessionScopedValue {
+    }
+
+    private static class ApiFilter implements Filter {
+        @Override
+        public void init(FilterConfig filterConfig) throws ServletException {
+        }
+
+        @Override
+        public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+                throws IOException, ServletException {
+            chain.doFilter(request, response);
+        }
+
+        @Override
+        public void destroy() {
+        }
+    }
+
+    private static final class AdminFilter extends ApiFilter {
+    }
+
+    private static final class HealthServlet extends HttpServlet {
+    }
+
+    private static final class AssetServlet extends HttpServlet {
+    }
+
+    private static final class NamedServlet extends HttpServlet {
+    }
+}

--- a/tests/src/com.google.inject.extensions/guice-servlet/4.0/user-code-filter.json
+++ b/tests/src/com.google.inject.extensions/guice-servlet/4.0/user-code-filter.json
@@ -1,0 +1,13 @@
+{
+  "rules" : [
+    {
+      "excludeClasses" : "**"
+    },
+    {
+      "includeClasses" : "com.google.inject.servlet.**"
+    },
+    {
+      "includeClasses" : "com_google_inject_extensions.guice_servlet.**"
+    }
+  ]
+}

